hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/BlockRecordManagerImpl.java

```diff
@@ -0,0 +1,397 @@
+     *
+     * @return the current block number, 0 of there is no blocks yet
+     */
+    @Override
+    public long lastBlockNo() {
+        return lastBlockInfo.lastBlockNumber();
+    }
+
+    /**
+     * Get the consensus time of the first transaction of the last block, this is the last completed immutable block.
+     *
+     * @return the consensus time of the first transaction of the last block, null if there was no previous block
+     */
+    @Nullable
+    @Override
+    public Instant firstConsTimeOfLastBlock() {
+        final var firstConsTimeOfLastBlock = lastBlockInfo.firstConsTimeOfLastBlock();
+        return firstConsTimeOfLastBlock != null
+                ? Instant.ofEpochSecond(firstConsTimeOfLastBlock.seconds(), firstConsTimeOfLastBlock.nanos())
+                : null;
+    }
+
+    /**
+     * Gets the hash of the last block
+     *
+     * @return the last block hash, null if no blocks have been created
+     */
+    @Nullable
+    @Override
+    public Bytes lastBlockHash() {
+        return getLastBlockHash(lastBlockInfo);
+    }
+
+    /**
+     * Returns the hash of the given block number, or {@code null} if unavailable.
+     *
+     * @param blockNo the block number of interest, must be within range of (current_block - 1) -> (current_block - 254)
+     * @return its hash, if available otherwise null
+     */
+    @Nullable
+    @Override
+    public Bytes blockHashByBlockNumber(final long blockNo) {
+        final Bytes blockHashes = lastBlockInfo.blockHashes();
+        final long blocksAvailable = blockHashes.length() / HASH_SIZE;
+        if (blockNo < 0) {
+            return null;
+        }
+        final long lastBlockNo = lastBlockInfo.lastBlockNumber();
+        final long firstAvailableBlockNo = lastBlockNo - blocksAvailable + 1;
+        // If blocksAvailable == 0, then firstAvailable == blockNo; and all numbers are
+        // either less than or greater than or equal to blockNo, so we return unavailable
+        if (blockNo < firstAvailableBlockNo || blockNo > lastBlockNo) {
+            return null;
+        } else {
+            long offset = (blockNo - firstAvailableBlockNo) * HASH_SIZE;
+            return blockHashes.slice(offset, HASH_SIZE);
+        }
+    }
+
+    // ========================================================================================================
+    // Private Methods
+
+    /**
+     * Get the block period from consensus timestamp. Based on
+     * {@link LinkedObjectStreamUtilities#getPeriod(Instant, long)} but updated to work on {@link Instant}.
+     *
+     * @param consensusTimestamp The consensus timestamp
+     * @return The block period from epoc the consensus timestamp is in
+     */
+    private long getBlockPeriod(@Nullable final Instant consensusTimestamp) {
+        if (consensusTimestamp == null) return 0;
+        final long nanos = consensusTimestamp.getEpochSecond() * 1_000_000_000L + consensusTimestamp.getNano();
```

Comment:
This comment made us stop and think. It turns out, this logic is more complicated that it needs to be, we can just have it be a simple integer division of `consensusTimestamp.getEpochSecond()` by `blockPeriodInSeconds`. We're just trying to figure out which block period we are in, where the first block period 0 is the first `blockPeriodInSeconds` seconds since the epoch, 1 is the second, and so on. I'll fix that.